### DISCUSSION TITLE
AULab Only Registers Virtual Audio Device Data When At Least One Output Device Is Physical

### DISCUSSION THREAD

**Discussion by rostislav-mitkov-peltekov** [ExistentialAudio/BlackHole#749] [Q&A]
First of all, thank you to the project for providing this source-code as open-source.
Secondly, the page [Running Multiple BlackHole Drivers](https://github.com/ExistentialAudio/BlackHole/wiki/Running-Multiple-BlackHole-Drivers) seems quite outdated. Resulted in much confusion and anxiety.
This may be unrelated to ExistentialAudio's BlackHole project, but I am at my wit's end and would love any hints/pointers/et cetera.

I. Clone, build, and install multiple drivers to use as virtual audio devices.
II. Install AULab and configure document to employ: 'bh-1_alpha' as input device; 'bh-2_beta' as output device.
III. Set system audio output to be: 'bh-1_alpha'. Observe audio meters registering _zero_ change: id est, greyed out — no change.
IV. Close and configure new document to employ: 'bh-1_alpha' as input device; 'Built-in Output' as Output.
V. Observe audio meters registering audio input and hear the audio playing through the physical output device 'Built-in Output'.

I've tried so many different configurations. Restarted so many times. Checked sample rates and clock source over and over. Triple checked all UID's were unique on every build. Even attempted building with operating system target of 12.4 to no avail.
At this point, I am leaning to the problem being on AULab's side. Although, perhaps new versions of BlackHole (I had been using previous versions in ~2022 with AULab no problem) break some AULab functionality. Routing audio between multiple instances of VLC (Video Lan Client) and BlackHole virtual audio devices is no problem though — provided one does not wish to specify channels through VLC (2ch only used), of course.

        — Very appreciative of your (the reader's) time, have a good one,
        — Ross

**Comment by rostislav-mitkov-peltekov** [comment]
Upon typing out the question and starting the discussion, I have realized that perhaps I could attempt to find an old commit of BlackHole from around a year ago to see if the drivers built with it produce the expected/desired behavior . . .

**Comment by devinroth** [comment]
Maybe try changing BlackHole endpoint type to USB instead of Virtual. That has been known to trick some apps into behaving properly.

**Reply by rostislav-mitkov-peltekov** [reply]
I very much appreciate the reply and suggestion. I also attempted building drivers from a previous commit around a year and a half ago; the exact same outcome as previously described resulted. This seems to be an issue entirely seperate from BlackHole and probably something solely to do with AULab/macOS.
As such, please excuse the unrelated discussion topic, but could I ask you how one changes the "BlackHole endpoint type to USB instead of Virtual"? I tried looking/searching for the keywords 'USB/usb/Virtual/virtual' (I doubt character-case matters) in 'Blackhole.c' — could not find any such attribute/parameter/argument/string/et cetera.
Once again, very much appreciate your time Mr. D. Roth. Good day.

**Comment by devinroth** [comment]
https://developer.apple.com/documentation/coreaudio/core_audio_enumerations/1494580-anonymous?changes=__8

But look them up in the CoreAudio headers. They have more info there.

Devin
